core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,202 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.controller
+
+import java.util.concurrent.locks.ReentrantLock
+import java.util.concurrent.{ArrayBlockingQueue, ConcurrentHashMap, CountDownLatch}
+
+import kafka.utils.CoreUtils.inLock
+import org.apache.zookeeper.AsyncCallback.{ACLCallback, Children2Callback, DataCallback, StatCallback, StringCallback, VoidCallback}
+import org.apache.zookeeper.Watcher.Event.EventType
+import org.apache.zookeeper.ZooKeeper.States
+import org.apache.zookeeper.data.{ACL, Stat}
+import org.apache.zookeeper.{CreateMode, WatchedEvent, Watcher, ZooKeeper}
+
+class ZookeeperClient(connectString: String, sessionTimeout: Int) {
+  private val initializationLock = new ReentrantLock()
+  private val isConnectedOrExpiredLock = new ReentrantLock()
+  private val isConnectedOrExpiredCondition = isConnectedOrExpiredLock.newCondition()
+  private val zNodeChangeHandlers = new ConcurrentHashMap[String, ZNodeChangeHandler]()
+  private val zNodeChildChangeHandlers = new ConcurrentHashMap[String, ZNodeChildChangeHandler]()
+  @volatile private var stateChangeHandlerOpt: Option[StateChangeHandler] = None
+  @volatile private var zooKeeper = new ZooKeeper(connectString, sessionTimeout, ZookeeperClientWatcher)
+  private val sessionContext = new ThreadLocal[ZooKeeper] {
+    override def initialValue(): ZooKeeper = zooKeeper
+  }
+
+  def handle(request: AsyncRequest): AsyncResponse = {
+    batch(Seq(request)).head
+  }
+
+  def batch(requests: Seq[AsyncRequest]): Seq[AsyncResponse] = {
+    import scala.collection.JavaConverters._
+    val countDownLatch = new CountDownLatch(requests.size)
+    val responseQueue = new ArrayBlockingQueue[AsyncResponse](requests.size)
+    requests.foreach {
+      case CreateRequest(path, data, acl, createMode, ctx) => sessionContext.get().create(path, data, acl.asJava, createMode, new StringCallback {
+        override def processResult(rc: Int, path: String, ctx: Any, name: String) = {
+          responseQueue.add(CreateResponse(rc, path, ctx, name))
+          countDownLatch.countDown()
+        }}, ctx)
+      case DeleteRequest(path, version, ctx) => sessionContext.get().delete(path, version, new VoidCallback {
+        override def processResult(rc: Int, path: String, ctx: Any) = {
+          responseQueue.add(DeleteResponse(rc, path, ctx))
+          countDownLatch.countDown()
+        }}, ctx)
+      case ExistsRequest(path, ctx) => sessionContext.get().exists(path, false, new StatCallback {
```

Comment:
Well, once the watcher fires, we need to reset the watcher again. It may not be convenient for the caller to explicitly call register again. Instead, it's more convenient if the watcher is set automatically on the read after the watcher fires, which is typically what the application has to do (This is what ZkClient does.).